### PR TITLE
feat(canvas)[OK-138] Add debouncing technique for actions on board  

### PR DESCRIPTION
**Pull request by adbista**
- Add debounce for object modification listener
- Add debounce in performAction
- Add debounce in toolbar
- create ToolbarContext to remove duplicated code
- Add debounce for undo and redo
- fix Bug with locking text object

### CODE REVIEW

File: client/src/components/board/toolbar-controls/ObjectSpecificControls.tsx [hunk 1]
```diff
@@ -1,65 +1,54 @@
...
-    };
-    socketEmitUpdateObject(socket, updateObjectData);
-
-    const objectId: string = modifiedObjectJSON._id;
-    const command = new ModifyCommand(canvas, clonedJSON, modifiedObjectJSON, objectId, handleStyleChange);
-    saveCommand(command);
-  };
+      setObjectStyle(canvas, modifiedObject, { [key]: newValue });
+      handleStyleChange();
+      //logger.log("Toolbar changing object style", key);
```
Comment (Strattoss) [inline client/src/components/board/toolbar-controls/ObjectSpecificControls.tsx]: Delete or uncomment

File: client/src/contexts/UndoRedoContext.tsx [hunk 2]
```diff
@@ -45,6 +47,8 @@ export const UndoRedoProvider: React.FC<{ children: React.ReactNode }> = ({ chil
     lastAction.undo(socket);
     redoStack.current.push(lastAction);
     canvas.renderAll();
+    //toast.dismiss();
+    //toast.message("Undo performed");  for testing purposes
```
Comment (Strattoss) [inline client/src/contexts/UndoRedoContext.tsx]: For testing purposes :)

File: client/src/contexts/UndoRedoContext.tsx [hunk 3]
```diff
@@ -61,11 +65,21 @@ export const UndoRedoProvider: React.FC<{ children: React.ReactNode }> = ({ chil
     lastAction.redo(socket);
     undoStack.current.push(lastAction);
     canvas.renderAll();
+    //toast.dismiss();
+    //toast.message("Redo performed"); for testing purposes
```
Comment (Strattoss) [inline client/src/contexts/UndoRedoContext.tsx]: For testing purposes

**Review by Strattoss — APPROVED** [review]
I'm truly impressed. You managed to weave the debounce code really elegantly into the existing code.

**Comment by Strattoss** [comment]
I have found a bug. It's easier to observe if debounce times are higher (for example 1 second).
If I select multiple (not one!) objects, move them and deselect them very quickly, the update is not sent to other users.
Here's an example of an update event being sent correctly:

https://github.com/user-attachments/assets/d0f6f431-bfbb-4822-9320-55f88d27bfa0

And here's the situation when the update event is not broadcasted at all to other users:

https://github.com/user-attachments/assets/5f2d6968-78cc-4aab-97a5-986bb28e7392



And I think I know where the problem sits.
We use debounce here:
obraz
But we check, what objects are selected here:
obraz

In the meantime, between the debounced `handleObjectModified` and `handleActiveSelectionModification`, the user can deselect objects or even choose different ones

**Comment by adbista** [comment]
> I have found a bug. It's easier to observe if debounce times are higher (for example 1 second). If I select multiple (not one!) objects, move them and deselect them very quickly, the update is not sent to other users. Here's an example of an update event being sent correctly:
> 
>  correct_update_sent.mp4 
> And here's the situation when the update event is not broadcasted at all to other users:
> 
>  no_update_sent.mp4 
> And I think I know where the problem sits. We use debounce here: obraz But we check, what objects are selected here: obraz
> 
> In the meantime, between the debounced `handleObjectModified` and `handleActiveSelectionModification`, the user can deselect objects or even choose different ones

@Strattoss  Could you please try it again? I think i fixed this issue, it was simple way, i just created two debounced functions: one for handleObjectModified and second for handleActiveSelectionModification.

**Comment by Adam0s007** [comment]
> > I have found a bug. It's easier to observe if debounce times are higher (for example 1 second). If I select multiple (not one!) objects, move them and deselect them very quickly, the update is not sent to other users. Here's an example of an update event being sent correctly:
> > correct_update_sent.mp4
> > And here's the situation when the update event is not broadcasted at all to other users:
> > no_update_sent.mp4
> > And I think I know where the problem sits. We use debounce here: obraz But we check, what objects are selected here: obraz
> > In the meantime, between the debounced `handleObjectModified` and `handleActiveSelectionModification`, the user can deselect objects or even choose different ones
> 
> @Strattoss  Could you please try it again? I think i fixed this issue, it was simple way, i just created two debounced functions: one for handleObjectModified and second for handleActiveSelectionModification.

**Comment by Strattoss** [comment]
> > I have found a bug. It's easier to observe if debounce times are higher (for example 1 second). If I select multiple (not one!) objects, move them and deselect them very quickly, the update is not sent to other users. Here's an example of an update event being sent correctly:
> > correct_update_sent.mp4
> > And here's the situation when the update event is not broadcasted at all to other users:
> > no_update_sent.mp4
> > And I think I know where the problem sits. We use debounce here: obraz But we check, what objects are selected here: obraz
> > In the meantime, between the debounced `handleObjectModified` and `handleActiveSelectionModification`, the user can deselect objects or even choose different ones
> 
> @Strattoss Could you please try it again? I think i fixed this issue, it was simple way, i just created two debounced functions: one for handleObjectModified and second for handleActiveSelectionModification.

It works correctly now

**Comment by Strattoss** [comment]
@Adam0s007 Is there any kind of debouncing for deleting objects? I seems like deleting happens instantaneously

**Comment by adbista** [comment]
> @Adam0s007 Is there any kind of debouncing for deleting objects? I seems like deleting happens instantaneously

Do we need debouncing for this? When we remove objects, they are no longer on the board, so there’s no way we can delete an object multiple times if it’s already removed
image
image

**Comment by Strattoss** [comment]
Oh, yes, you're right

**Pull request merged**